### PR TITLE
Integrate sbt-bioconda into biopet plugin

### PR DESCRIPTION
**Pull request by rhpvorderman**
TODO:
- [x] Release sbt-bioconda

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -13,7 +13,7 @@ node('local') {
         }
 
         stage('Build & Test') {
-            sh "${tool name: 'sbt 0.13.15', type: 'org.jvnet.hudson.plugins.SbtPluginBuilder$SbtInstallation'}/bin/sbt -no-colors clean scalafmt scripted headerCheck"
+            sh "${tool name: 'sbt 0.13.15', type: 'org.jvnet.hudson.plugins.SbtPluginBuilder$SbtInstallation'}/bin/sbt -no-colors clean scalafmt test headerCheck"
```
Comment (ffinfo) [inline Jenkinsfile]: Please update sbt here to 1.0.4

**Review by ffinfo — CHANGES_REQUESTED** [review]

File: build.sbt [hunk 2]
```diff
@@ -74,7 +74,8 @@ releaseProcess := Seq[ReleaseStep](
   commitNextVersion,
   pushChanges
 )
-
+libraryDependencies += "org.testng" % "testng" % "6.14.2" % "test"
```
Comment (ffinfo) [inline build.sbt]: It's better practice to use `Test` instead of `"test"`

File: src/main/scala/nl/biopet/sbtbiopet/BiopetUtils.scala [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package nl.biopet.sbtbiopet
```
Comment (ffinfo) [inline src/main/scala/nl/biopet/sbtbiopet/BiopetUtils.scala:22]: Is this is not something to move to common-utils?

File: src/main/scala/nl/biopet/sbtbiopet/BiopetUtils.scala [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+  def markdownExtractChapter(markdown: String,
+                             chapter: String,
+                             includeHeader: Boolean = true): String = {
+    // Regex is anchored with ^pattern$ by default.
+    // Matches any number of #'s, a whitespace (tab or space) that does not have to exist and the chapter.
+    val chapterRegex =
+      s"(#+)([\\t ]*)($chapter)".r("hashtags", "whitespace", "heading")
+    val headingDepth = chapterRegex.findFirstMatchIn(markdown) match {
+      case Some(r) => r.group("hashtags").length
+      case None    => throw new Exception(s"Chapter not found: $chapter")
```
Comment (ffinfo) [inline src/main/scala/nl/biopet/sbtbiopet/BiopetUtils.scala]: to generic Exception

File: src/main/scala/nl/biopet/sbtbiopet/BiopetUtils.scala [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    * @return The chapter including header and subheaders as a string.
+    */
+  def markdownExtractChapter(markdown: String,
+                             chapter: String,
+                             includeHeader: Boolean = true): String = {
+    // Regex is anchored with ^pattern$ by default.
+    // Matches any number of #'s, a whitespace (tab or space) that does not have to exist and the chapter.
+    val chapterRegex =
+      s"(#+)([\\t ]*)($chapter)".r("hashtags", "whitespace", "heading")
+    val headingDepth = chapterRegex.findFirstMatchIn(markdown) match {
```
Comment (ffinfo) [inline src/main/scala/nl/biopet/sbtbiopet/BiopetUtils.scala]: Just a suggestion, not required but maybe nicer:
```scala
chapterRegex.findFirstMatchIn(markdown)
  .map(_.group("hashtags").length)
  .getOrElse(throw new IllegalStateException(s"Chapter not found: $chapter"))
```

File: src/main/scala/nl/biopet/sbtbiopet/BiopetUtils.scala [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+      case None    => throw new Exception(s"Chapter not found: $chapter")
+    }
+
+    // General matcher that returns true if the line is
+    // a. a markdown chapter (starting with #)
+    // b. Starting with a maximum of headingDepth hashtags.
+    // If the maximum dept is exceeded (another # was found) then retuns false.
+    // This way we can split the markdown at a desired depth.
+    def matcher(string: String): Boolean = {
+      val regex = s"^(#{1,$headingDepth})[^#]([\\t ]*)(.*)$$".r
```
Comment (ffinfo) [inline src/main/scala/nl/biopet/sbtbiopet/BiopetUtils.scala]: it's better to put the regex outside this method as a `private lazy val`. This way the regex is only created once.

Comment (ffinfo) [reply src/main/scala/nl/biopet/sbtbiopet/BiopetUtils.scala]: this is true for all regexes

File: src/main/scala/nl/biopet/sbtbiopet/BiopetUtils.scala [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+      regex.findFirstMatchIn(string).isDefined
+    }
+
+    val text = markdown.split(lineSeparator).toList
+    val chapters = splitStringList(text, matcher)
+    val correctChapter = chapters.find(lines => {
```
Comment (ffinfo) [inline src/main/scala/nl/biopet/sbtbiopet/BiopetUtils.scala]: I think what you try to do here is this:
```scala
    chapters.find(lines => lines.headOption
      .exists(chapterRegex.findFirstMatchIn(_).isDefined))
      .map(lines => if (includeHeader) lines else lines.drop(1))
      .map(_.mkString(lineSeparator))
      .getOrElse(throw new IllegalStateException(
        "Cannot validate correct chapter after parsing. " +
          "Please contact application maintainers."))
```

File: src/main/scala/nl/biopet/sbtbiopet/BiopetUtils.scala [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+  /**
+    * Split a list of strings at a certain element.
+    * @param stringList The list of strings
+    * @param splitter Function that determines whether the list should be split at this line.
+    * @return A list of lists of string.
+    */
+  def splitStringList(stringList: List[String],
+                      splitter: String => Boolean): List[List[String]] = {
+    val buffers = stringList.foldLeft(ListBuffer[ListBuffer[String]]()) {
+      case (result, line) =>
```
Comment (ffinfo) [inline src/main/scala/nl/biopet/sbtbiopet/BiopetUtils.scala]: Double cases can sometimes be resolved into 1 like this:
```scala
    stringList.foldLeft(ListBuffer[ListBuffer[String]]()) {
      case (result, line) if splitter(line) || result.lastOption.isEmpty => 
        result += ListBuffer(line)
      case (result, line) =>
        result.last += line
        result
    }.map(buffer => buffer.toList).toList
```

Also if possible try to not use a val if it's not needed.

**Review by ffinfo — CHANGES_REQUESTED** [review]

File: src/main/scala/nl/biopet/sbtbiopet/BiopetUtils.scala (hunk 3 above)
Comment (rhpvorderman) [reply src/main/scala/nl/biopet/sbtbiopet/BiopetUtils.scala:22]: Yes. Let's do this for the next release of common-utils and sbt-biopet.

**Review by ffinfo — APPROVED** [review]

**Pull request merged**